legate/core/runtime.py

```diff
@@ -711,14 +729,389 @@ def record_partition(self, index_space, functor, index_partition):
+        """
+        This is a class containing a list of constraints for fusing ops
+        It emits whether or not a given list of ops can be fused
+        """
+        self.constraints = []
+        self.ops = ops
+        self.contexts = contexts
+        self.runtime=runtime
+        self.partitioners = []
+        self.strategies = []
+
+    def register_constraint(self, fusion_constraint_rule):
+        self.constraints.append(fusion_constraint_rule)
+
+    def supress_small_fusions(self, intervals, threshold):
+        #find if there's a fusable sub window of length
+        #greater than or equal to fusion_thresh
+        final_set = []
+        fusable=False
+        for interval in intervals:
+            if interval[1] - interval[0]  >=threshold:
+                final_set.append(interval)
+                fusable = True
+            else:
+                for i in range(interval[0], interval[1]):
+                    final_set.append((i, i+1))
+        return fusable, final_set
+
+    def can_fuse(self):
+        #starttime = datetime.datetime.now()
+        for op in reversed(self.ops):
+            must_be_single = any(len(gop.scalar_outputs) > 0 for gop in [op])
+            partitioner = Partitioner(self.runtime, [op], must_be_single=must_be_single)
+            self.partitioners.append( partitioner )
+            strategy = partitioner.partition_stores()
+            for output, part, in zip(op._outputs, op._output_parts):
+                partition = strategy.get_partition(part)
+                output.set_key_partition(partition)
+                key_part = partition
+                for input in op._inputs:
+                    if input.shape==output.shape:
+                        input.set_key_partition(key_part)
+            #if len(op.inputs)>1:
+            #    proj = strategy.get_projection(op._input_parts[1])
+            self.strategies.append(strategy)
+        self.strategies.reverse()
+        """
+        stoptime = datetime.datetime.now()
+        delta=stoptime-starttime
+        total = delta.total_seconds() * 1000.0
+        print("partime", total, len(self.ops))
+        """     
+        streamApply = False
+        #starttime = datetime.datetime.now()
+        if not streamApply:
+            results = [constraint.apply(self.contexts, self.runtime, self.ops, self.partitioners, self.strategies) for constraint in self.constraints]
+            all_fusable = [result[0] for result in results]
+            interval_sets = [result[1] for result in results]
+        else:
+            alpha = self.constraints[0].apply(self.contexts, self.runtime, self.ops, self.partitioners, self.strategies)
+            beta = self.constraints[1].apply2(self.contexts, self.runtime, self.ops, alpha[1],self.partitioners, self.strategies)
+            beta = self.constraints[2].apply2(self.contexts, self.runtime, self.ops, beta[1], self.partitioners, self.strategies)
+        #print("beta", beta)
+        """
```

Comment:
yes this is messy, will clean up in the next day or so. The imporant thing is that the `apply2` class of methods are much more efficient at generating the windows of fusable ops than the `apply` methods

Reply:
`apply2` takes a set of fusable windows (generated by another constraint), and generates an updated set of fusable windows using the current constraint. the old `apply` simply generates a set of fusable windows for this constraint only (the sets of fusable windows for each constraint then have to be coalesced, which is not very efficient)